minoc/src/main/java/org/opencadc/minoc/ArtifactAction.java

```diff
@@ -191,21 +202,103 @@ void parsePath() {
         log.debug("authToken: " + authToken);
     }
     
+    Artifact getArtifact(URI artifactURI, ArtifactDAO dao) throws ResourceNotFoundException {
+        Artifact artifact = dao.get(artifactURI);
+        if (artifact == null) {
+            throw new ResourceNotFoundException("not found: " + artifactURI);
+        }
+        if (artifact.storageLocation == null) {
+            throw new ResourceNotFoundException("not available: " + artifactURI);
+        }
+        return artifact;
+    }
+    
     /**
      * Create a valid artifact uri.
      * @param uri The input string.
      * @return The artifact uri objecdt.
      */
     private URI createArtifactURI(String uri) {
         try {
+            log.debug("artifactURI: " + uri);
             artifactURI = new URI(uri);
             InventoryUtil.validateArtifactURI(ArtifactAction.class, artifactURI);
             return artifactURI;
         } catch (URISyntaxException e) {
-            String message = "Illegal artifact URI: " + uri;
+            String message = "illegal artifact URI: " + uri;
             log.debug(message, e);
             throw new IllegalArgumentException(message);
         }
     }
+    
+    protected StorageAdapter getStorageAdapter() {
+        // lazy init
+        if (storage == null) {
+            PropertiesReader pr = new PropertiesReader("minoc.properties");
+            String adapterKey = StorageAdapter.class.getName();
+            String adapterClass = pr.getFirstPropertyValue(adapterKey);
+            if (adapterClass == null) {
+                throw new IllegalStateException("no storage adapter specified in minoc.properties");
+            }
+            try {
+                Class c = Class.forName(adapterClass);
+                Object o = c.newInstance();
+                StorageAdapter sa = (StorageAdapter) o;
+                log.debug("StorageAdapter: " + sa);
+                storage = sa;
+                return sa;
+            } catch (Throwable t) {
+                throw new IllegalStateException("failed to load storage adapter: " + adapterClass, t);
+            }
+        }
+        return storage;
+    }
+    
+    protected ArtifactDAO getArtifactDAO() {
+        // lazy init
+        if (artifactDAO == null) {
+            ArtifactDAO newDAO = new ArtifactDAO();
+            newDAO.setConfig(getDaoConfig());
+            artifactDAO = newDAO;
+        }
+        return artifactDAO;
+    }
+    
+    protected DeletedEventDAO getDeletedEventDAO(ArtifactDAO src) {
+        // lazy init
+        if (deletedEventDAO == null) {
+            DeletedEventDAO newDAO = new DeletedEventDAO(src);
+            deletedEventDAO = newDAO;
+        }
+        return deletedEventDAO;
+    }
+    
+    static Map<String, Object> getDaoConfig() {
+        Map<String, Object> config = new HashMap<String, Object>();
+        PropertiesReader pr = new PropertiesReader("minoc.properties");
```

Comment:
The minoc.properties file is read twice -- should read once (maybe lazy init of a "PropertiesReader props" member variable?)

Reply:
And document the required props in catalina.properties to configure the pool in context.xml

Reply:
minoc.properties is read once per request (whether it's from availability or one of the actions).  The other dao just uses the copy constructor.

Reply:
I wasn't clear - it is read once to get the StorageAdapter impl  and once to get the SQLGenerator impl and database info.

On minoc.properties:

Technically, knowing the database name is not required since in most cases is it in the JDBC url (catalina.properties). Putting it into the InitDatabase (ctor) means it is used to qualify the table name... SQLGenerator currently ignores the database setting and I'll only add support for putting it into the SQL if needed. In most cases: configuring it in two places only adds an opportunity for a mistake. So that should probably be left out for now and a null database name passed into the init. I'll probably change InitDatabase to ignore the database until both it and SQLGenerator need it.

Reply:
Re: reading config--since this pull request I've changed it so that the MultiValuedProperties are obtained once from the file and used from memory to get all config items.

Okay, I'll remove database and make SQL generator optional.